src/draco/io/ply_decoder.cc

```diff
@@ -136,6 +139,65 @@ Status PlyDecoder::DecodeFaceData(const PlyElement *face_element) {
   return OkStatus();
 }
 
+Status PlyDecoder::DecodeTexCoordData(const PlyElement *face_element) {
+  if (face_element == nullptr) {
+    return Status(Status::INVALID_PARAMETER, "face_element is null");
+  }
+  const PlyProperty *texture_coordinates =
+      face_element->GetPropertyByName("texcoord");
+  if (texture_coordinates == nullptr || !texture_coordinates->is_list()) {
+    return OkStatus();
+  }
+
+  // Allocate attribute for texture coordinates.
+  // There is one pair of texture coordinates per triangle corner.
+  const int num_corners = out_mesh_->num_faces() * 3;
+  
+  
+  std::unique_ptr<PointAttribute> attr(new PointAttribute());
+  attr->Init(GeometryAttribute::TEX_COORD, 2, DT_FLOAT32, false, num_corners);
+    
+  PlyPropertyReader<float> uv_reader(texture_coordinates);
+  AttributeValueIndex corner_index(0);
+  
+  auto pushTexcoordPair = [uv_reader, &corner_index, &attr]
+      (uint64_t offset, uint64_t index) -> void {
+    float uv_value[2];
+    uv_value[0] = uv_reader.ReadValue(static_cast<int>(offset + index * 2));
+    uv_value[1] = uv_reader.ReadValue(static_cast<int>(offset + index * 2 + 1));
+    attr->SetAttributeValue(corner_index, uv_value);
+    corner_index++;
+  };
+
+  const int64_t num_polygons = face_element->num_entries();
+  for (int i = 0; i < num_polygons; ++i) {
+    const int64_t uv_list_offset = texture_coordinates->GetListEntryOffset(i);
+    const int64_t uv_list_size = texture_coordinates->GetListEntryNumValues(i);
+    if (uv_list_size < 6 || uv_list_size % 2 != 0) {
+      continue;  // Need at least three pairs of UV coordinates per polygon.
+    }
+
+    // Triangulate polygon assuming the polygon is convex.
+    const int64_t num_triangles = uv_list_size / 2 - 2;
+    pushTexcoordPair(uv_list_offset, 0);
+    for (int64_t ti = 0; ti < num_triangles; ++ti) {
+      for (int64_t c = 1; c < 3; ++c) {
+        pushTexcoordPair(uv_list_offset, ti + c);
```

Comment:
I don't think this would add one value for each corner, not to mention the corner index seems to be wrong.

Basically you would have to add value for each corner where the corner index is (3 * triangle_index + c) where c={0, 1, 2}

Reply:
Good catch, fixed. This moves through vertices in the same order as in `DecodeFaceData`. When triangulating a polygon with more than 3 sides, the first vertex is repeated in each triangle.

Reply:
I understand that the first vertex is repeated but even so, each triangle of the polygon will have a distinct corner attached to the "first" vertex. E.g. if you have a quad there will be two corners attached to it. In the current implementation only the corner from the first triangle is going to be processed which is wrong because it messes up the way how Draco maps corners to triangles + it leaves some corners with uninitialized values. Again, as an example for a quad, we would expect to have two triangles with corners: (0, 1, 2), (3, 4, 5) where they would be mapped to values (0, 1, 2), (0, 2, 3).

Reply:
Just to make sure, are you looking at the change from the latest commit? https://github.com/google/draco/pull/1028/commits/eee2335ac31b51ae08647b383bcf03064107d74a

```
for (int64_t ti = 0; ti < num_triangles; ++ti) {
  pushTexcoordPair(uv_list_offset, 0);
  for (int64_t c = 1; c < 3; ++c) {
    pushTexcoordPair(uv_list_offset, ti + c);
  }
}
```

For a quad with corners (0, 1, 2, 3), this should push 6 pairs of UVs in the order (0, 1, 2, 0, 2, 3). Is that not correct?

Reply:
Just to make sure, are you looking at the change from the latest commit? https://github.com/google/draco/pull/1028/commits/eee2335ac31b51ae08647b383bcf03064107d74a

```
for (int64_t ti = 0; ti < num_triangles; ++ti) {
  pushTexcoordPair(uv_list_offset, 0);
  for (int64_t c = 1; c < 3; ++c) {
    pushTexcoordPair(uv_list_offset, ti + c);
  }
}
```

For a quad with corners (0, 1, 2, 3), this should push 6 pairs of UVs in the order (0, 1, 2, 0, 2, 3). Is that not correct?

Reply:
Sorry about that. Not sure what I was looking at, you are right.

One other potential issue I can see is that, if I understand it correctly, the DecodeTexCoordData() is called before we decode any vertex data. I think this may cause some issues because AddAttributeWithConnectivity assumes that we have some valid points in the mesh already. Can you try to move this decoding after DecodeVertexData() to see if it helps?